core/routing_spray.go

```diff
@@ -0,0 +1,313 @@
+	bundleData map[string]sprayMetaData
+}
+
+// NewBinarySpray creates new instance of BinarySpray
+func NewBinarySpray(c *Core) BinarySpray {
+	log.WithFields(log.Fields{
+		"L": sprayl,
+	}).Debug("Initialised BinarySpray")
+	// register our custom metadata-block
+	extensionBlockManager := bundle.GetExtensionBlockManager()
+	if !extensionBlockManager.IsKnown(ExtBlockTypeBinarySprayBlock) {
+		// since we already checked if the block type exists, this really shouldn't ever fail...
+		_ = extensionBlockManager.Register(NewBinarySprayBlock(0))
+	}
+	return BinarySpray{
+		c:          c,
+		bundleData: make(map[string]sprayMetaData),
+	}
+}
+
+// NotifyIncoming tells the routing algorithm about new bundles.
+// In this case, we check, whether we are the originator of this bundle
+// If yes, then we initialise the remaining Copies to L
+// If not we attempt to ready the routing-metadata-block end get the remaining copies
+func (bs BinarySpray) NotifyIncoming(bp BundlePack) {
+	if metadataBlock, err := bp.Bundle.ExtensionBlock(ExtBlockTypeBinarySprayBlock); err == nil {
+		binarySprayBlock := metadataBlock.Value.(*BinarySprayBlock)
+		metadata := sprayMetaData{
+			sent:            make([]bundle.EndpointID, 0),
+			remainingCopies: binarySprayBlock.RemainingCopies(),
+		}
+
+		// if the bundle has a PreviousNodeBlock, add it to the list of nodes which we know to have the bundle
+		if pnBlock, err := bp.Bundle.ExtensionBlock(bundle.ExtBlockTypePreviousNodeBlock); err == nil {
+			metadata.sent = append(metadata.sent, pnBlock.Value.(*bundle.PreviousNodeBlock).Endpoint())
+		}
+
+		bs.bundleData[bp.ID()] = metadata
+		log.WithFields(log.Fields{
+			"bundle": bp.ID(),
+		}).Debug("SprayAndWait received bundle from foreign host")
+	} else {
+		metadata := sprayMetaData{
+			sent:            make([]bundle.EndpointID, 0),
+			remainingCopies: sprayl,
+		}
+		bs.bundleData[bp.ID()] = metadata
+		log.WithFields(log.Fields{
+			"bundle": bp.ID(),
+		}).Debug("SprayAndWait initialised new bundle from this host")
+	}
+}
+
+// SenderForBundle returns the Core's ConvergenceSenders.
+// If a node has more than 1 copy left it will send floor(copies/2) to the peer
+// and keep roof(copies/2) for itself
+func (bs BinarySpray) SenderForBundle(bp BundlePack) (css []cla.ConvergenceSender, del bool) {
+	metadata, _ := bs.bundleData[bp.ID()]
+	// if there are no copies left, we just wait until we meet the recipient
+	if metadata.remainingCopies < 2 {
+		log.WithFields(log.Fields{
+			"bundle": bp.ID(),
+		}).Debug("Not relaying bundle because there are copies left")
+		return nil, false
+	}
+
+	for _, cs := range bs.c.convergenceSenders {
+		var skip = false
+		for _, eid := range metadata.sent {
+			if cs.GetPeerEndpointID() == eid {
+				skip = true
+				break
+			}
+		}
+
+		if !skip {
+			css = append(css, cs)
+			metadata.sent = append(metadata.sent, cs.GetPeerEndpointID())
+
+			// we send half our remaining copies
+			sendCopies := metadata.remainingCopies / 2
```

Comment:
Should we check at this point whether the counter is already zero?

Reply:
We only send to a single peer at at time